rfcs/20240315-algorithm-support.md

```diff
@@ -0,0 +1,166 @@
+# [RFC] Add algorithm to dot_general's parameters in the StableHLO specification
+
+Status: Review<br/>
+Initial version: 03/15/2024<br/>
+Last updated: 03/15/2024<br/>
+Discussion thread: [GitHub](https://github.com/openxla/stablehlo/pull/2096)
+
+## Motivation
+
+We would like to allow explicitly selecting the algorithm used for individual
+`dot_general` instructions, enforcing a well-defined numeric precision. We think
+that the current precision values (`DEFAULT`, `HIGH`, `HIGHEST`) are not
+specific enough for this, because they mean different things between hardware
+platforms and even between GPU types.
+
+We also want to make the API flexible enough to support additional algorithms
+that might be faster on certain hardware
+([6xBF16](https://arxiv.org/pdf/1904.06376.pdf), 3xTF32, etc) that can't fit
+into the limited precision_config values of `DEFAULT`, `HIGH` and `HIGHEST`. For
+example "6xBF16" provides similar precision to F32, but about 2x performance on
+some GPUs.
+
+When an algorithm is not supported by an accelerator, the program should fail
```

Comment:
Maybe I miss the nuances and I apologize if I do. I feel this choice impacts fungibility. A vendor may want to totally specialize for a specific algorithm but a customer may want to run an unmodified JAX on HW from vendors A and B.   Maybe there should be a way for vendors (or JAX) to specify "fall-back" types, sort of the reverse mapping from special type to 'DEFAULT', 'HIGH' and 'HIGHEST'? (I wonder why we don't offer 'LOW' and 'LOWEST' ;-)

Reply:
I think that JAX's request was that on the XLA side we should require exact algorithms.
But on the JAX side, they could write some fallback logic using the current GPU/TPU type, etc.
I thought that it's not needed to specify the fallback in the stableHLO, JAX could just decide the final algorithm before writing the stableHLO.

Maybe @hawkinsp can comment on this?
(I think he did in some other forum, but I can't find it anymore.)

Reply:
Yes. If I specify an exact algorithm, I want that exact algorithm. Otherwise I get, e.g., unexpected precision changes when moving code between platforms. I'd much rather I get an explicit error.

But I agree it makes sense to also allow less-specific choices as well.

Reply:
I'll note that this was specifically requested by one of our internal users, who found that the numerics of their computation completely changed when moving from TPU to GPU or vice versa, but the problem was almost entirely resolved when using the same matmul algorithm on both.

Reply:
Shouldn't we allow to control the behavior with a flag. Default could be something like "strict" vs "permissive" numerical accuracy, something akin to the -ffast-math option, except for data types?

Reply:
Do you think this flag should be in XLA or JAX?

Reply:
The problem with a flag is that it is global and thus does not nicely.

Reply:
It also gets a bit weird if you want to go from jax2tf. I think it needs to be serialized.

Reply:
Does the user need to define that less specific is ok, or the compilation pipeline? This "fallback" notion could exist as an mlir pass if the pipeline can determine it